### PR TITLE
[WIP]: initial commit of ripley in numpy-oriented style

### PR DESCRIPTION
**Pull request by ljwolf**
This is done to facilitate discussion about a simpler `scipy.stats`-style API for Ripley functions. If this is felt more reasonable in `pointpats`, that's totally cool. Hopefully we might incorporate this style of logic into other areas of esda, amiing at `scipy.stats` as the style of implementation.

My main exploration with this is to satisfy 4 principles:

0. Be as performant as possible. This means *use numpy and scipy as often as possible. avoid list comprehensions and for loops.* I think this is achieved since there is only one for loop/comprehension in the module.
1. Keep time-to-estimation short. This means *custom classes and secondary/tertiary instantions of classes should be minimized*. I think this is satisfied because all estimators take a numpy array and return numpy arrays or result namedtuples. 
2. Make a user of `scipy.stats` feel immediately at home. This means *return values should be `namedtuples` with standard names for their properties*, such as `statistic` and `pvalue`, regardless of the statistical result. I think this is satisfied, since the APIs, naming, and structure are very similar.
3. Use keywords, not classes for power features. This means *opening up sklearn-style distance and metric specifications that allow users to specify their own nulls*. I think this is satisfied, since power users can simulate their own processes, compute any fancy distance metric they want, and send them along to the statistics using the `distance` keyword. It also means all of these can accept *network* distances, and support `spaghetti`. 
 
I will document this further & fill out with testing in the coming days. I aim to have a fully documented & working example by next PySAL development call (1 May 2020).

### CODE REVIEW

File: esda/ripley.py [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+            upper_tri_distances = distances[numpy.triu_indices_from(distances, k=1)]
+        else:
+            raise ValueError(
+                f"Shape of inputted distances is not square, nor is the upper triangular"
+                " matrix matching the number of input points. The shape of the input matrix"
+                " is {distance.shape}, but required shape is ({upper_tri_n},) or ({n},{n})"
+            )
+    n_pairs_less_than_d = (upper_tri_distances < support.reshape(-1, 1)).sum(axis=1)
+    intensity = n / _area(hull)
+    k_estimate = n_pairs_less_than_d * 2 / intensity
```
Comment (ljwolf) [inline esda/ripley.py:461]: This is missing a divide by n, I believe.

File: esda/ripley.py [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+    hull = _prepare_hull(tree.data)  # and this over replications
+    core_kwargs["hull"] = hull
+
+    observed_support, observed_statistic = stat_function(tree, **core_kwargs)
+    core_kwargs["support"] = observed_support
+
+    if keep_replications:
+        replications = numpy.empty((len(observed_support), n_replications))
+    for i_replication in range(n_replications):
+        replications_i = stat_function(tree, **core_kwargs)[1]
```
Comment (ljwolf) [inline esda/ripley.py:543]: This is not currently doing replications.

**Comment by jGaboardi** [comment]
I added [this issue](https://hackmd.io/AnnMZhUARq6qnizL8ElWoA?both#Ripley-in-numpy-oriented-style-Levi) the agenda of 2020-05-01 meeting (next week).

**Comment by ljwolf** [comment]
Hmm. no, I don't know why that'd fail... nothing in this touches join counts. On my own local computer, I get failures in esda/master on these tests using libpysal/master (which, I assume these allowed failures are testing?)

**Pull request closed without merging**